File: 0100-same-tree/0100-same-tree.cpp
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == NULL && q == NULL) return true;
        if (p == NULL || q == NULL || p->val != q-> val) return false;
        return isSameTree(p->right, q->right) && isSameTree(p->left, q->left);
    }
};

File: 0100-same-tree/0100-same-tree.py
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        def check(n, m):
            if (not n and m) or (n and not m): return False
            if not n and not m: return True
            if n.val != m.val: return False

            if n.left and m.left:
                chk = check(n.left, m.left)
                if not chk: return False
            if n.right and m.right:
                chk = check(n.right, m.right)
                if not chk: return False
            if ((n.left and not m.left) or
                (n.right and not m.right) or
                (not n.left and m.left) or
                (not n.right and m.right)):
                return False

            return True

        return check(p, q)